botocore/errorfactory.py

```diff
@@ -0,0 +1,38 @@
+import logging
+import threading
+
+from botocore.utils import CachedProperty
+from botocore.exceptions import ClientError
+
+logger = logging.getLogger(__name__)
+
+
+class ServiceErrorFactory:
+    ClientError = ClientError
+
+    def __init__(self, service_model):
+        self._service_model = service_model
+        self._lock = threading.Lock()
+
+    @CachedProperty
+    def _error_shapes(self):
+        shapes = {}
+        for op_name in self._service_model.operation_names:
+            op_model = self._service_model.operation_model(op_name)
+            for shape in op_model.error_shapes:
+                shapes[shape.name] = shape
+        return shapes
+
+    def __getattr__(self, attr):
+        if attr.startswith("_"):
```

Comment:
What was the reasoning for adding this validation? I do realize that underscores are typically reserved for private api, but just curious.